### DISCUSSION TITLE
760 folders of dependencies is too many

### DISCUSSION THREAD

**Discussion by manake** [josdejong/mathjs#2785] [Q&A]
1. 760 folders of dependencies is too many for such small and simple project. Simplifying things is recommended.
    - Perhaps this library can be simplified to close to 0 dependencies.
    - Especially `webpack` or `babel` and such seem to be redundant.
2. Without building the /src/ folder is missing `mathjs/entry/pureFunctionsAny.generated.js` and a few other files so it can't be imported easily in a browser as ESM module.
    - I think this should be a part of /src/ by default and no building should be required.
3. After building /src/dependenciesAny/ and /src/dependenciesNumber/ have 500+ files total with nothing but references in them. That shouldn't be the case.

**Overall, rethinking the project structure and simplifying would be recommended as it's too complex and too messy currently.**

**I highly recommend using projects like https://www.npmjs.com/package/three as reference, which has both 0 dependencies and instantly working /src/ and correct /dist/** (overall, it's 10/10 experience to setup).

**Comment by josdejong** [comment]
@manake I hope you realize your tone is quite belittling and patronizing, saying things like "... such small and simple project" and "I highly recommend ...". Please change your tone to something more constructive.

From this question and and your other questions I strongly get the feeling that you're quite new to JavaScript, node.js, npm, and this project. That is fine. I'm open to any question. But please don't try to give advice if you're not familiar with the matter. First ask questions and try to understand what's going on and why things are as they are right now. The entry files like in dependenciesAny are there with a good reason. The dependencies like `typed-function`, `complex.js`, and `decimal.js`, are there with a reason. Dev dependencies like `webpack` and `babel` are there with a reason. Etcetera. 

You can indeed not consume the library straight from the source code in `/src`, but all you have to do is run `npm install && npm run build` and then refer to the generated output in `/lib`.

As for the actual topic you touch: yes, it would be nice to simplify the the project. There are a couple of big discussions in this regard, like #2741 and #2742. 

So, one question to you: why are you checking out the source code instead of using the npm package? Are you planning on contributing to the project?

**Comment by manake** [comment]
I'm sorry, I'm not a native English speaker. I meant to just state facts and not "belittle" or anything like that.

On the surface this looked to me like an extension to `Math` with many isolated small functions.

```
all you have to do is run npm install && npm run build and then refer to the generated output in /lib.
```
I installed via `npm install mathjs` and I tried to load like `import * as mathjs from 'lib/esm/index.js'` and it doesn't work because it says that `@babel/...` references must start with `./` or `../`. So, so far I gave up on trying to `import` into browser. Perhaps `importmap` would help but it's not supported by many browsers. I'm currently sticking to loading plain math.min.js.

**Reply by josdejong** [reply]
> I'm sorry, I'm not a native English speaker. I meant to just state facts and not "belittle" or anything like that.

👍 

> On the surface this looked to me like an extension to Math with many isolated small functions.

It's _much_ more. Just read the description, feature list, and examples to get a feel: https://mathjs.org/#an-extensive-math-library-for-javascript-and-nodejs

> I installed via `npm install mathjs` and I tried to load like `import * as mathjs from 'lib/esm/index.js'` and it doesn't work because it says that `@babel/...` references must start with `./` or `../`. So, so far I gave up on trying to `import` into browser. Perhaps `importmap` would help but it's not supported by many browsers. I'm currently sticking to loading plain math.min.js.

Unfortunately, the ES modules cannot be used straight in the browser right now because there are dependencies in the code which need to be resolved from `node_modules`, and some of the dependencies are CommonJS and not yet ES modules. ES modules cannot deal with commonjs dependencies and must point to a real relative path, so npm imports like `import typed from 'typed-function'` do not work in the browser whilst they work fine in node.js. 

When working in the browser you have to use the bundle, as explained in the getting started: https://mathjs.org/docs/getting_started.html. Most common is to create a web application using a framework like React, with Webpack and Babel and everything under the hood, which will solve all of the hassle for you so you can just use ES modules and any npm dependency.
